### PR TITLE
Fixes #1537 Distribution Config Export Commands

### LINKED ISSUE #1537: Provide way to re-export Quickstart managed config settings
## Motivation
We need a way to quickly/easily re-export the configuration entities managed by Quickstart on-demand so that we can avoid configuration drift and make transitions between module and Drupal core versions easier.

## Proposed Resolution
Provide a way to automate the bulk-updating of our exported configuration files.

### Describe the solution you'd like
A custom drush command might be the best long-term solution.  But any solution that makes bulk-updating our exported configuration files easier would be welcome.

### Describe alternatives you've considered
One disadvantage of not using the features module for managing our Quickstart-provided configuration is that we don't have access to the tools built-in to the features module to automate updating exported feature modules.  We could look into using the APIs or drush commands (or potentially even just some of the code) provided by the features module.

### PR DESCRIPTION
**Pull request by tadean**
## Description
One task frequently during development is the re-export of module owned configuration to snapshot changes made in the development environment. In particular this would be desirable after the transition to D10 to catch configuration drift. This PR introduces functionality that examines module-owned configuration items and attempts to re-export changes in the active configuration to the module files.

A command is also added to output a single configuration item with the site-specific hashes removed, since Drupal console is no longer available in our D10 build. (The drush equivalent does not remove the site-specific hashes)

- `drush az-core-distribution-config`
- `drush az-core-config-export-single`

These are development tools to assist with exporting distribution configuration.

**NOTE: our configuration contains a few edge cases that are not currently handled by this PR but I think would make good future improvements in functionality:**

- Our permission list in roles contain all applicable permissions. Missing active permissions will be removed when re-export. Ideally, this functionality should merge the inactive permissions so they are not lost - i.e. It would be better if only available permissions that were explicitly removed were removed in the export. Take care when re-exporting roles.
- New dependent config is not export, only files already present. It would be nice if dependent configuration was export if it's not already in another module file.
- Quickstart override config will be re-export to the non-overridden version, e.g. metatag-enabled settings. It would be good if this functionality only exported to the overridden copy e.g. `az_seo`

I hopefully intend to handle the above cases in future improvements to this initial functionality.

## Related issues
#1537 

## How to test

- `lando drush az-core-config-export-single configuration.item.name`, e.g. `lando drush az-core-config-export-single cas.settings`
- verify the chosen configuration is output to the console
- `lando drush az-core-distribution-config` or `lando drush az-core-distribution-config az_module1,az_module2,az_module3` with a comma delimited list of module names
- step through the process of confirming which items should be export
- verify that any changes were written to the module config folders

See description for cases which are not currently handled by this initial command (further work or future PRs).

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by tadean** [comment]
Example output:
```
$ lando drush -y az-core-distribution-config az_cas,az_publication,az_event
az_cas...
    -- unmodified -- az_cas.settings
    -- unmodified -- cas.settings
az_event...
    -- unmodified -- core.base_field_override.node.az_event.promote
    -- unmodified -- core.date_format.az_event_day
    -- unmodified -- core.date_format.az_event_month

 //     Update [az_event/config/install] core.entity_form_display.node.az_event.default?: yes.                          

 //     Update [az_event/config/install] core.entity_view_display.node.az_event.az_card?: yes.                          

 //     Update [az_event/config/install] core.entity_view_display.node.az_event.az_card_alternate?: yes.                

 //     Update [az_event/config/install] core.entity_view_display.node.az_event.az_row?: yes.                           

 //     Update [az_event/config/install] core.entity_view_display.node.az_event.default?: yes.                          

 //     Update [az_event/config/install] core.entity_view_display.node.az_event.teaser?: yes.                           

    -- unmodified -- core.entity_view_display.taxonomy_term.az_event_categories.az_badge
    -- unmodified -- field.field.node.az_event.field_az_attachments
    -- unmodified -- field.field.node.az_event.field_az_body
    -- unmodified -- field.field.node.az_event.field_az_contacts
    -- unmodified -- field.field.node.az_event.field_az_event_category
    -- unmodified -- field.field.node.az_event.field_az_event_date
    -- unmodified -- field.field.node.az_event.field_az_featured_image_as_thumb
    -- unmodified -- field.field.node.az_event.field_az_link
    -- unmodified -- field.field.node.az_event.field_az_location
    -- unmodified -- field.field.node.az_event.field_az_media_thumbnail_image
    -- unmodified -- field.field.node.az_event.field_az_photos
    -- unmodified -- field.field.node.az_event.field_az_subheading
    -- unmodified -- field.field.node.az_event.field_az_summary
    -- unmodified -- field.field.node.az_event.field_az_trellis_id
    -- unmodified -- field.storage.node.field_az_event_category
    -- unmodified -- field.storage.node.field_az_event_date
    -- unmodified -- field.storage.node.field_az_trellis_id
    -- unmodified -- node.type.az_event
    -- unmodified -- pathauto.pattern.az_event_url_pattern
    -- unmodified -- taxonomy.vocabulary.az_event_categories
    -- unmodified -- views.view.az_events
    -- unmodified -- field.field.node.az_event.field_az_metatag
az_publication not installed.
```

```
$ git status

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   modules/custom/az_event/config/install/core.entity_form_display.node.az_event.default.yml
	modified:   modules/custom/az_event/config/install/core.entity_view_display.node.az_event.az_card.yml
	modified:   modules/custom/az_event/config/install/core.entity_view_display.node.az_event.az_card_alternate.yml
	modified:   modules/custom/az_event/config/install/core.entity_view_display.node.az_event.az_row.yml
	modified:   modules/custom/az_event/config/install/core.entity_view_display.node.az_event.default.yml
	modified:   modules/custom/az_event/config/install/core.entity_view_display.node.az_event.teaser.yml
```

**Comment by tadean** [comment]
We discussed the idea of adding this functionality to a new proposed development tools module, perhaps as a dependency of our dev metapackage (not necessarily as part of `az_quickstart`. This PR shouldn't be merged yet, but is available for testing (or dev use of the command).

File: modules/custom/az_core/src/Commands/AZCoreConfigCommands.php [hunk 1]
```diff
@@ -101,4 +136,104 @@ public function addMissingPermissions() {
...
+          $original = $storage->read($item);
+          $active = $this->configFactory->get($item)->get();
+          // Remove site-specific identifiers.
+          unset($original['_core']);
+          unset($original['uuid']);
+          unset($active['_core']);
+          unset($active['uuid']);
+          // Diff the state of configuration to check for changes.
+          $diff = $this->configDiffer->diff($original, $active);
+          if (!$diff->isEmpty()) {
```
Comment (trackleft) [inline modules/custom/az_core/src/Commands/AZCoreConfigCommands.php:195]:
```
 ------ ----------------------------------------------------------------- 
8  Line   modules/custom/az_core/src/Commands/AZCoreConfigCommands.php     
9 ------ ----------------------------------------------------------------- 
10  195    Call to deprecated method isEmpty() of class                     
11         Drupal\Component\Diff\Diff:                                      
12         in drupal:10.1.0 and is removed from drupal:11.0.0. There is no  
13           replacement.                                                   
14 ------ ----------------------------------------------------------------- 
```

Comment (trackleft) [reply modules/custom/az_core/src/Commands/AZCoreConfigCommands.php:195]: See https://www.drupal.org/node/3337942

Comment (trackleft) [inline modules/custom/az_core/src/Commands/AZCoreConfigCommands.php:195]:
```suggestion
          if (!empty($diff)) {
```

**Comment by tadean** [comment]
Added a `prepareRoleConfig()` method to the config export that transparently prevents removal of permissions that aren't currently active, making role export more generally useful for managing permissions.

**Comment by joeparsons** [comment]
Very excited for this!

**Comment by tadean** [comment]
- Added ignoring export of overridden configuration
- Added draft beginning of dependency discovery export

**Comment by tadean** [comment]
- Began refactor to chained dependency resolution
- Added initial functionality to ignore certain changes or revert dev site defaults so they do not end up in exported configuration

**Comment by tadean** [comment]
Brought up to date with main.

**Comment by tadean** [comment]
Added bidirectional dependency resolution, e.g. configuration from the site is identified as potentially export appropriate if it depends on elements of the distro, or elements of the distro depend on it. E.g. in the case of having added to the dev site a new field to person, and a new view that displays persons:

```
Examining dependencies...
 Export NEW dependent configuration field.field.node.az_person.field_a_new_field? (yes/no) [yes]:
 > no

 Export NEW dependent configuration field.storage.node.field_a_new_field? (yes/no) [yes]:
 > no

 Export NEW dependent configuration views.view.a_new_view? (yes/no) [yes]:
 > no
```

**Review by joeparsons — APPROVED** [review]

**Pull request merged**